File: Tutorials/Cracking the Coding Interview/Data Structure/Trie.cs
/*
Tries: Contacts
https://www.hackerrank.com/challenges/ctci-contacts/problem
*/
using System;
using System.Collections;

namespace Trie
{
    class Program
    {
        private class TrieNode
        {
            private Hashtable hashtable;
            private int count = 0;
            public TrieNode(){hashtable = new Hashtable();}
            public int Count
            {
                get{return count;}
            }
            public object this[object key]
            {
                get{return hashtable[key];}
                set{hashtable[key]=value;}
            }
            public void Add(object key, object value)
            {
                hashtable.Add(key, value);
            }
            public bool ContainsKey(object key){return hashtable.ContainsKey(key);}
            public void PrintItems(string prefix="")
            {
                if(ContainsKey('\0'))
                    Console.WriteLine(prefix);
                foreach(var key in hashtable.Keys)
                {
                    if((char)key=='\0')
                        continue;
                    ((TrieNode)hashtable[key]).PrintItems(prefix + (char)key);
                }
            }
            public void IncreaseCount()
            {
                count += 1;
            }
        }
        class Trie
        {
            TrieNode root;
            
            public Trie(){root = new TrieNode();}
            public void Add(string str)
            {
                int index = 0;
                TrieNode node = root;
                while(index<str.Length && node.ContainsKey(str[index]))
                {
                    node = (TrieNode)node[str[index++]];
                }
                if(index < str.Length)
                {
                    do
                    {
                        node.Add(str[index], new TrieNode());
                        node = (TrieNode)node[str[index++]];
                    }while(index < str.Length);
                    node.Add('\0', null);
                    IncreaseCount(str);
                }
            }
            public int CountAllStartsWith(string prefix="")
            {
                int index = 0;
                TrieNode node = root;
                while(index<prefix.Length && node.ContainsKey(prefix[index]))
                {
                    node = (TrieNode)node[prefix[index++]];
                }
                if(index < prefix.Length)
                    return 0;
                return node.Count;
            }
            public void PrintItems(string prefix="")
            {
                root.PrintItems(prefix);
            }
            private void IncreaseCount(string str)
            {
                int index = 0;
                TrieNode node = root;
                while(index<str.Length)
                {
                    node.IncreaseCount();
                    node = (TrieNode)node[str[index]];
                    index += 1;
                }
                node.IncreaseCount();
            }
        }
        static void Main(string[] args)
        {
            int numberOfInstructions = Convert.ToInt32(Console.ReadLine());
            Trie trie = new Trie();
            for(int index_instruction = 0; index_instruction < numberOfInstructions; index_instruction++){
                string[] tokens_op = Console.ReadLine().Split(' ');
                string op = tokens_op[0];
                string contact = tokens_op[1];
                switch(op)
                {
                    case "add":
                        trie.Add(contact);
                        break;
                    case "find":
                        Console.WriteLine(trie.CountAllStartsWith(contact));
                        break;
                }
            }
        }
    }
}

File: Tutorials/Cracking the Coding Interview/Data Structure/Trie.cpp
/*
Tries: Contacts
https://www.hackerrank.com/challenges/ctci-contacts/problem

The knowledge about the trie is acquired from ``演算法筆記 - String"
url: http://www.csie.ntnu.edu.tw/~u91029/String.html
*/
#include<iostream>
#include<map>
#include<string>

using namespace std;

class Trie
{
private:
    typedef struct TrieNode
    {
        map<char, TrieNode*> pNextLetter;
        int count;
        bool word;
        TrieNode()
        {
            count = 0;
            word = false;
        }
        ~TrieNode()
        {
            for(auto it = pNextLetter.begin(); it!=pNextLetter.end(); ++it)
                if(it->second) delete it->second;
        }
    } TrieNode;
public:
    void add(string str)
    {
        TrieNode *pNode = root;
        for(const char *s = str.c_str(); *s; s+=1)
        {
            if(! pNode->pNextLetter[*s])
                pNode->pNextLetter[*s] = new TrieNode;
            pNode = pNode->pNextLetter[*s];
        }

        if(pNode->word)
            return;
        pNode->word = true;

        pNode = root;
        for(const char *s = str.c_str(); *s; s+=1)
        {
            pNode->count += 1;
            pNode = pNode->pNextLetter[*s];
        }
        pNode->count += 1;
    }
    int countPrefix(string str)
    {
        TrieNode *pNode = root;
        const char *s = str.c_str();
        for(; *s; s+=1)
        {
            if(! pNode->pNextLetter[*s])
                return 0;
            pNode = pNode->pNextLetter[*s];
        }
        return *s ? 0 : pNode->count;
    }
    bool search(string str)
    {
        TrieNode *pNode = root;
        for(const char *s = str.c_str(); *s; s+=1)
        {
            if(! pNode->pNextLetter[*s])
                return false;
            pNode = pNode->pNextLetter[*s];
        }
        return pNode->word;
    }
    void traversal()
    {
        string str;
        traversal(root, str);
    }
private:
    TrieNode *root = new TrieNode;
    void traversal(TrieNode *pNode, string buffer)
    {
        if(!pNode) return;
        if(pNode->word)
            cout << buffer << "\n";
        buffer.resize(buffer.size()+1);
        for(auto it=pNode->pNextLetter.begin(); it!=pNode->pNextLetter.end(); ++it)
        {
            buffer.back() = it->first;
            traversal(pNode->pNextLetter[it->first], buffer);
        }
    }
};
int main()
{
    int n;
    string op;
    string contact;

    Trie contacts;
    
    cin >> n;
    for(int a0 = 0; a0 < n; a0++)
    {
        cin >> op >> contact;
        if(op.compare("add")==0)
            contacts.add(contact);
        if(op.compare("find")==0)
            cout << contacts.countPrefix(contact) << "\n";
    }
    return 0;
}
